### PR TITLE
[MERGE][IMP] event[_sale]: improve event views (frontend/backend) and ticket sale flow

### PR DESCRIPTION
**Pull request by pko-odoo**
PURPOSE

Improve global event usability by adding a graph view as well as various
changes on the ticket sale flow.

We also improved the track proposal layout, added a handy button to invite
partners to your event and prepared some data for the event_barcode module.

SPECIFICATIONS

- Add a "Invite" button on the event, that allows sending invitations though
  the mass_mailing / mass_sms features
- Add a graph view for the event.event model
- Allow styling the description and the biography when sending a track proposal
  with basic plaintext2html usage
- Improve event and sale views integration in the event_sale module
  (stat buttons and next activity generation)
- General improvements to the "auto confirm" flow of registrations
- Prepare data for the event_barcode module, to make it "event agnostic"

More details in underlying commits

LINKS

Task ID : 2093336
Enterprise PR: odoo/enterprise#7760

### CODE REVIEW

File: addons/website_event_track/controllers/main.py [hunk 1]
```diff
@@ -146,11 +147,12 @@ def event_track_proposal_post(self, event, **post):
...
             'partner_email': post['email_from'],
             'partner_phone': post['phone'],
-            'partner_biography': escape(post['biography']),
+            'partner_biography': plaintext2html(post['biography']),
             'event_id': event.id,
             'tag_ids': [(6, 0, tags)],
             'user_id': False,
-            'description': escape(post['description'])
+            'description': plaintext2html(post['description']),
+            'image': base64.b64encode(post['image'].read())
```
Comment (awa-odoo) [inline addons/website_event_track/controllers/main.py]: The tester will probably tell you that as well but it raises an error if you don't select an image (and the field is not mandatory).

File: addons/website_event_track/views/event_track_templates.xml [hunk 2]
```diff
@@ -441,8 +447,8 @@
         <li><b>Proposed By</b>: <t t-esc="track.partner_name or track.partner_email"/></li>
         <li><b>Mail</b>: <a t-attf-href="mailto:#{track.partner_email}"><t t-esc="track.partner_email"/></a></li>
         <li><b>Phone</b>: <t t-esc="track.partner_phone"/></li>
-        <li><b>Speakers Biography</b>: <t t-esc="track.partner_biography"/></li>
-        <li><b>Talk Introduction</b>: <t t-esc="track.description"/></li>
+        <li><b>Speakers Biography</b>: <t t-raw="track.partner_biography"/></li>
+        <li><b>Talk Introduction</b>: <t t-raw="track.description"/></li>
```
Comment (awa-odoo) [inline addons/website_event_track/views/event_track_templates.xml]: Be extra careful with "t-raw", even though I think it's ok for this case.

File: addons/website_event_track/views/event_track_templates.xml [hunk 3]
```diff
@@ -376,6 +376,12 @@
                                     <textarea  class="form-control" name="biography" rows="5"/>
                                 </div>
                             </div>
+                            <div class="form-group row">
```
Comment (awa-odoo) [inline addons/website_event_track/views/event_track_templates.xml]: Side note: The name of the tab in my browser is "event_track_proposal".
Could be nice if we can do something about it.

Selection_357

File: addons/website_event_track/views/event_track_views.xml [hunk 4]
```diff
@@ -118,12 +118,13 @@
...
                         </div>
-                        <widget name="web_ribbon" title="Archived" bg_color="bg-danger" attrs="{'invisible': [('active', '=', True)]}"/>
-                        <field name="kanban_state" widget="state_selection"/>
+                        <widget name="web_ribbon" text="Archived" bg_color="bg-danger" attrs="{'invisible': [('active', '=', True)]}"/>
+                        <field name="image" widget="image" class="oe_avatar"/>
                         <div class="oe_title">
                             <label for="name" class="oe_edit_only"/>
                             <h1>
                                 <field name="name" placeholder="e.g. Inspiring Business Talk"/>
+                                <field name="kanban_state" widget="state_selection" class="ml-3"/>
```
Comment (awa-odoo) [inline addons/website_event_track/views/event_track_views.xml]: Are we sure about moving that button there?
Didn't see anything about it in the task pad.

File: addons/website_event_track/controllers/main.py (hunk 1 above)
Comment (pko-odoo) [reply addons/website_event_track/controllers/main.py]: I didn't get any error when submitting the form without an image

File: addons/event_sale/models/sale_order.py [hunk 5]
```diff
@@ -22,6 +24,18 @@ def action_confirm(self):
...
+        return {
+            "name": _("Attendees"),
+            "type": "ir.actions.act_window",
+            "res_model": "event.registration",
+            "view_mode": "tree",
+            "domain": [('sale_order_id', '=', self.id)]
+        }
+
+    def _compute_attendee_count(self):
+        self.attendee_count = len(self.env['event.registration'].search([('sale_order_line_id', 'in', self.order_line.ids)]))
```
Comment (tde-banana-odoo) [inline addons/event_sale/models/sale_order.py]: Beware computed fields should work in batch. You should probably use a read_group (see existing code, there are a lot of them in odoo :) ) .

File: addons/event/data/email_template_data.xml [hunk 6]
```diff
@@ -421,5 +421,26 @@
             <field name="user_signature" eval="False"/>
         </record>
 
+        <record id="event_invitation" model="mail.template">
```
Comment (tde-banana-odoo) [inline addons/event/data/email_template_data.xml]: According to me the whole "contact" without mass mailing makes no sense. I mean, why are we doing that for event and not for other applications ? If people need a way to do this, they can define a template and add it in context action of res.partner . To check with PO but according to me it is unnecessary code to maintain for a feature people should be able to do in custo.

The mass mailing behavior makes sense as it is quite fairly integrated into odoo so I think that one can be kept.

The other way round would be to define a small event.invite wizard that in the background creates a composer and send mails.

Side note: if you don't have website_event installed I bet you will ran into troubles :)

File: addons/event_sale/models/event_registration.py [hunk 7]
```diff
@@ -31,13 +32,13 @@ def _check_ticket_seats_limit(self):
             if record.event_ticket_id.seats_max and record.event_ticket_id.seats_available < 0:
                 raise ValidationError(_('No more available seats for this ticket'))
 
-    def _check_auto_confirmation(self):
```
Comment (tde-banana-odoo) [inline addons/event_sale/models/event_registration.py]: Why removing this ?

File: addons/event_sale/models/event_registration.py [hunk 8]
```diff
@@ -48,6 +49,26 @@ def create(self, vals):
...
+    def write(self, values):
+        if 'event_ticket_id' in values:
+            new_ticket_id = self.env['event.event.ticket'].browse(values['event_ticket_id'])
+            if self.event_ticket_id != new_ticket_id:
+                self._sale_order_ticket_type_change_notify(self, new_ticket_id)
+
+        result = super(EventRegistration, self).write(values)
+        return result
+
+    def _sale_order_ticket_type_change_notify(self, event_registration, new_ticket_id):
```
Comment (tde-banana-odoo) [inline addons/event_sale/models/event_registration.py]: I don't really understand, people change ticket on a registration and we trigger an activity on the SO, but don't do anything else ? What kind of bad behavior is that ?

File: addons/event_sale/models/sale_order.py [hunk 9]
```diff
@@ -22,6 +24,18 @@ def action_confirm(self):
                     .for_xml_id('event_sale', 'action_sale_order_event_registration')
         return res
 
+    def action_view_attendee_list(self):
+        return {
+            "name": _("Attendees"),
+            "type": "ir.actions.act_window",
+            "res_model": "event.registration",
+            "view_mode": "tree",
+            "domain": [('sale_order_id', '=', self.id)]
```
Comment (tde-banana-odoo) [inline addons/event_sale/models/sale_order.py]: in self.ids so that it works in batch.

File: addons/website_event_track/models/event_track.py [hunk 10]
```diff
@@ -66,14 +66,14 @@ def _get_default_stage_id(self):
...
-    location_id = fields.Many2one('event.track.location', 'Room')
+    location_id = fields.Many2one('event.track.location', 'Location')
     event_id = fields.Many2one('event.event', 'Event', required=True)
     color = fields.Integer('Color', related="stage_id.color")
     priority = fields.Selection([
         ('0', 'Low'), ('1', 'Medium'),
         ('2', 'High'), ('3', 'Highest')],
         'Priority', required=True, default='1')
-    image = fields.Image("Image", related='partner_id.image_128', store=True, readonly=False)
+    image = fields.Image("Image", required=True)
```
Comment (tde-banana-odoo) [inline addons/website_event_track/models/event_track.py]: I would not set an image as required ...

File: addons/website_event_track/views/event_track_templates.xml (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/website_event_track/views/event_track_templates.xml]: Why using t-raw ? isn't it possible to use t-field ?

File: addons/event_sale/models/event_registration.py (hunk 7 above)
Comment (pko-odoo) [reply addons/event_sale/models/event_registration.py]: When the auto_confirm flag is set on the event the registration should always be set on confirmed

File: addons/event_sale/models/event_registration.py (hunk 8 above)
Comment (pko-odoo) [reply addons/event_sale/models/event_registration.py]: The idea here is only to notify the responsible. He will then decide himself what to do with the sale order.

File: addons/event/models/event_event.py [hunk 11]
```diff
@@ -229,6 +230,22 @@ def _compute_date_end_tz(self):
             else:
                 event.date_end_located = False
 
+    @api.depends('date_begin', 'date_end')
+    def _compute_is_ongoing(self):
+        for event in self:
+            event.is_ongoing = event.date_begin <= fields.Datetime.now() < event.date_end
+
+    def _search_is_ongoing(self, operator, value):
+        if (operator not in ['=', '!=']) or (value not in [True, False]):
```
Comment (dbeguin) [inline addons/event/models/event_event.py]: not isinstance(value, bool)

File: addons/website_event_track/views/event_track_templates.xml [hunk 12]
```diff
@@ -338,7 +339,7 @@
                             <div class="form-group row">
                                 <label class="col-lg-3 col-md-4 col-form-label" for="description">Talk Introduction</label>
                                 <div class="col-lg-7 col-md-8">
-                                    <textarea  class="form-control" name="description" style="min-height: 120px"/>
+                                    <textarea  class="form-control o_wysiwyg_loader" name="description" style="min-height: 120px"/>
```
Comment (dbeguin) [inline addons/website_event_track/views/event_track_templates.xml]: inline style ! berk

File: addons/event_sale/models/event_registration.py [hunk 13]
```diff
@@ -10,6 +10,7 @@ class EventRegistration(models.Model):
     _inherit = 'event.registration'
 
     event_ticket_id = fields.Many2one('event.event.ticket', string='Event Ticket', tracking=True)
+    is_paid = fields.Boolean('Is Paid', default=False)
```
Comment (dbeguin) [inline addons/event_sale/models/event_registration.py]: Awa:

> 'default false ça sert a rien'

File: addons/event_sale/models/event_registration.py [hunk 14]
```diff
@@ -93,6 +86,9 @@ def _prepare_attendee_values(self, registration):
             })
         return att_data
 
+    def mark_as_paid(self):
```
Comment (dbeguin) [inline addons/event_sale/models/event_registration.py]: awa:

> Et pourquoi c'est pas privé ?

File: addons/event/models/event_event.py [hunk 15]
```diff
@@ -229,6 +231,22 @@ def _compute_date_end_tz(self):
             else:
                 event.date_end_located = False
 
+    @api.depends('date_begin', 'date_end')
+    def _compute_is_ongoing(self):
+        for event in self:
+            event.is_ongoing = event.date_begin <= fields.Datetime.now() < event.date_end
+
+    def _search_is_ongoing(self, operator, value):
```
Comment (tde-banana-odoo) [inline addons/event/models/event_event.py]: Would be cool if you have some spare time to write at least one test for this one :) .

File: addons/event/models/event_event.py [hunk 16]
```diff
@@ -229,6 +231,22 @@ def _compute_date_end_tz(self):
...
+        if (operator not in ['=', '!=']) or (not isinstance(value, bool)):
+            return []
```
Comment (tde-banana-odoo) [inline addons/event/models/event_event.py]: As it should not happen for a boolean field, you could simply raise a ValueError.

File: addons/event_sale/models/event_registration.py (hunk 7 above)
Comment (tde-banana-odoo) [reply addons/event_sale/models/event_registration.py]: Ok.

File: addons/event_sale/models/event_registration.py (hunk 8 above)
Comment (tde-banana-odoo) [reply addons/event_sale/models/event_registration.py]: Ok

File: addons/event_sale/models/event_registration.py [hunk 17]
```diff
@@ -48,6 +49,26 @@ def create(self, vals):
...
+        result = super(EventRegistration, self).write(values)
+        return result
+
+    def _sale_order_ticket_type_change_notify(self, event_registration, new_ticket_id):
+        render_context = {
+            'registration': event_registration,
+            'old_ticket': event_registration.event_ticket_id.name,
+            'new_ticket': new_ticket_id.name
+        }
+        event_registration.sale_order_id.activity_schedule_with_view('mail.mail_activity_data_warning',
```
Comment (tde-banana-odoo) [inline addons/event_sale/models/event_registration.py]: You could simply make a regular method on registration model, which will lead to
```
for registration in self:
    registration.sale_order_id.activity_schedule() ...
```
It is easier to read that way :) .

File: addons/event_sale/models/event_registration.py [hunk 18]
```diff
@@ -48,6 +49,26 @@ def create(self, vals):
...
+             user_id=event_registration.event_id.user_id.id or self.env.ref("base.user_root").id,
```
Comment (tde-banana-odoo) [inline addons/event_sale/models/event_registration.py]: User root is not a real user. I think you could check for
1/ event user_id;
2/ order user_id;
3/ current user if not public;
4/ admin;

File: addons/event_sale/models/event_registration.py [hunk 19]
```diff
@@ -65,20 +86,20 @@ def _prepare_attendee_values(self, registration):
             })
         return att_data
 
-    def summary(self):
-        res = super(EventRegistration, self).summary()
-        if self.event_ticket_id.product_id.image_128:
-            res['image'] = '/web/image/product.product/%s/image_128' % self.event_ticket_id.product_id.id
-        information = res.setdefault('information', {})
-        information.append((_('Name'), self.name))
+    def _mark_as_paid(self):
```
Comment (tde-banana-odoo) [inline addons/event_sale/models/event_registration.py]: _action_set_paid ? That way we are synchronized with other action method on registration model.

File: addons/website_event_track/models/event_track.py [hunk 20]
```diff
@@ -66,14 +66,14 @@ def _get_default_stage_id(self):
...
-    location_id = fields.Many2one('event.track.location', 'Room')
+    location_id = fields.Many2one('event.track.location', 'Location')
     event_id = fields.Many2one('event.event', 'Event', required=True)
     color = fields.Integer('Color', related="stage_id.color")
     priority = fields.Selection([
         ('0', 'Low'), ('1', 'Medium'),
         ('2', 'High'), ('3', 'Highest')],
         'Priority', required=True, default='1')
-    image = fields.Image("Image", related='partner_id.image_128', store=True, readonly=False)
+    image = fields.Image("Image")
```
Comment (tde-banana-odoo) [inline addons/website_event_track/models/event_track.py]: maybe a max width / max height ? I don't think we should keep too large images, what do you think ?

File: addons/event/views/event_views.xml [hunk 21]
```diff
@@ -620,6 +629,8 @@
                 <search string="Event Registration">
                     <field name="id" string="Registration ID"/>
                     <field name="name" string="Participant" filter_domain="['|', '|', ('name', 'ilike', self), ('email', 'ilike', self), ('origin', 'ilike', self)]"/>
+                    <filter string="Ongoing Events" name="ongoing_events" domain="[('event_id.is_ongoing', '=', True)]"/>
```
Comment (tde-banana-odoo) [inline addons/event/views/event_views.xml]: I would name it filter_is_ongoing. This convention is not enforced but we try to name filter with name_<field_name>, easier to find when searching. To propagate to enterprise PR if you change :) .

File: addons/event/models/event_registration.py [hunk 22]
```diff
@@ -249,6 +249,14 @@ def get_date_range_str(self):
...
+    def registration_summary(self):
         self.ensure_one()
-        return {'information': []}
+        return {
+            'registration': {
+                'id': self.id,
+                'partner_id': self.partner_id.id,
+                'state': self.state,
+                'name': self.name or _('Attendee'),
+                'information': [(_('Event'), self.event_id.name), (_('Name'), self.name)]
```
Comment (awa-odoo) [inline addons/event/models/event_registration.py]: Could be simplified to return a dict (and not a dict within a "registration" key).

Also, the "information" key seems very random.
You already have the name (under the "name" key).
Could simply add a "event_name" key, and handle the rest at display ?

File: addons/event/views/event_views.xml [hunk 23]
```diff
@@ -227,8 +227,8 @@
                         <widget name="web_ribbon" text="Archived" bg_color="bg-danger" attrs="{'invisible': [('active', '=', True)]}"/>
                         <field name="kanban_state" widget="state_selection" class="ml-auto"/>
                         <div class="oe_title">
-                            <label for="name" class="oe_edit_only"/>
-                            <h1><field name="name" placeholder="Event Name"/></h1>
+                            <label for="name" class="oe_edit_only" string="Event Name"/>
+                            <h1><field name="name" placeholder="e.g. Conference for Architects"/></h1>
```
Comment (awa-odoo) [inline addons/event/views/event_views.xml]: Seems unrelated to this commit's purpose.

File: addons/event_sale/models/event_registration.py [hunk 24]
```diff
@@ -60,20 +60,17 @@ def _synchronize_so_line_values(self, so_line):
...
-    def summary(self):
-        res = super(EventRegistration, self).summary()
-        if self.event_ticket_id.product_id.image_128:
-            res['image'] = '/web/image/product.product/%s/image_128' % self.event_ticket_id.product_id.id
-        information = res.setdefault('information', {})
-        information.append((_('Name'), self.name))
+    def registration_summary(self):
+        res = super(EventRegistration, self).registration_summary()
+        information = res.get('registration').setdefault('information', {})
         information.append((_('Ticket'), self.event_ticket_id.name or _('None')))
```
Comment (awa-odoo) [inline addons/event_sale/models/event_registration.py]: Same here.
Could append a "ticket_name" key and handle the rest at display level.

Backend should mostly be "display agnostic" and return base information.
The frontend takes care of displaying these information in an appropriate manner.

File: addons/website_event_track/static/src/js/website_event_track_proposal.js [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+var publicWidget = require('web.public.widget');
+var weDefaultOptions = require('web_editor.wysiwyg.default_options');
+var wysiwygLoader = require('web_editor.loader');
+
+publicWidget.registry.websiteEventTrackProposal = publicWidget.Widget.extend({
+    selector: '.o_website_event_talk_proposal',
+
+    start: function() {
+        var self = this;
+        _.each($('textarea.o_wysiwyg_loader'), function (textarea) {
```
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/website_event_track_proposal.js]: 2 things here.

1.
You should place yourself after the "start" to make sure everything is properly loaded.
`return this._super.apply(this, arguments).then(function() { ... });`
Or better yet extract to another function to make it more readable:
`return this._super.apply(this, arguments).then(this.loadWysiwyg.bind(this));`

2.
Avoid global selectors whenever possible.
Use `this.$('textarea.o_wysiwyg_loader');`

File: addons/website_event_track/static/src/js/website_event_track_proposal.js [hunk 26]
```diff
@@ -0,0 +1,36 @@
+odoo.define('website_event_track.website_event_track_proposal', function (require) {
+'use strict';
+
+var publicWidget = require('web.public.widget');
+var weDefaultOptions = require('web_editor.wysiwyg.default_options');
+var wysiwygLoader = require('web_editor.loader');
+
+publicWidget.registry.websiteEventTrackProposal = publicWidget.Widget.extend({
```
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/website_event_track_proposal.js]:
```
/**
* This small widget is used to initialize wysiwyg editors on the talk proposal page.
* This allows more layout flexibility when writing your description / biography / ...
*/
```

File: addons/website_event_track/static/src/js/website_event_track_proposal.js [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+            ];
+
+            var options = {
+                height: 200,
+                minHeight: 80,
+                toolbar: toolbar,
+                styleWithSpan: false,
+                styleTags: _.without(weDefaultOptions.styleTags, 'h1', 'h2', 'h3'),
+            };
+            wysiwygLoader.load(self, $textarea[0], options);
```
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/website_event_track_proposal.js]: No need to extract to a variable.
Remove $textarea and textarea from constructor.
Just use `$(this)[0]` here.

File: addons/website_event_track/static/src/js/website_event_track_proposal.js [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+        });
+        return this._super.apply(this, arguments);
+    },
+
+});
+});
```
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/website_event_track_proposal.js]: `return publicWidget.registry.websiteEventTrackProposal;`

To make it overridable by other modules if needed.

File: addons/event/models/event_event.py [hunk 29]
```diff
@@ -170,12 +170,13 @@ def _get_default_stage_id(self):
     date_end = fields.Datetime(string='End Date', required=True, tracking=True)
```
Comment (awa-odoo) [inline addons/event/models/event_event.py]: Also, in this commit: "[IMP] event: adds graph view and adds change for event_barcode"

You should include a link to the enterprise PR for reference:

odoo/enterprise#7760

(Not 100% sure about syntax, check if github understands that).

File: addons/mass_mailing_event/models/event.py [hunk 30]
```diff
@@ -20,3 +20,16 @@ def action_mass_mailing_attendees(self):
...
+            name=_('Mass Mail Invitation'),
+            type='ir.actions.act_window',
+            res_model='mailing.mailing',
+            view_mode='form',
+            target='current',
+            context=dict(
+                default_mailing_model_id=self.env.ref('base.model_res_partner').id,
+            ),
+        )
+        return mass_mailing_action
```
Comment (awa-odoo) [inline addons/mass_mailing_event/models/event.py]: Strange syntax:

```
return {
    'name': _('Mass Mail Invitation')
    ...
}
```

File: addons/mass_mailing_event_sms/models/event.py [hunk 31]
```diff
@@ -12,3 +12,8 @@ def action_mass_mailing_attendees(self):
         action = super(Event, self).action_mass_mailing_attendees()
         action['view_id'] = self.env.ref('mass_mailing_sms.mailing_mailing_view_form_mixed').id
         return action
+
+    def action_invite_contacts(self):
+        action = super(Event, self).action_invite_contacts()
```
Comment (awa-odoo) [inline addons/mass_mailing_event_sms/models/event.py]: Could add the same small comment as method above.

File: addons/mass_mailing_event/models/event.py [hunk 32]
```diff
@@ -20,3 +20,16 @@ def action_mass_mailing_attendees(self):
             ),
         )
         return mass_mailing_action
+
+    def action_invite_contacts(self):
+        mass_mailing_action = dict(
+            name=_('Mass Mail Invitation'),
```
Comment (awa-odoo) [inline addons/mass_mailing_event/models/event.py]: The name key should not be translated.

File: addons/event_sale/models/event_registration.py [hunk 33]
```diff
@@ -46,6 +54,12 @@ def write(self, vals):
                 self.env['sale.order.line'].browse(vals['sale_order_line_id'])
             )
             vals.update(so_line_vals)
+
+        if vals.get('event_ticket_id'):
+            new_ticket_id = self.env['event.event.ticket'].browse(vals['event_ticket_id'])
```
Comment (awa-odoo) [inline addons/event_sale/models/event_registration.py]: Could test the ids before browsing the model to avoid unnecessary requests.

```
if self.event_ticket_id and self.event_ticket_id.id != vals['event_ticket_id']:
    self._sale_order_ticket_type_change_notify(self.env['event.event.ticket'].browse(vals['event_ticket_id']))
```

File: addons/event_sale/models/event_registration.py [hunk 34]
```diff
@@ -60,6 +74,19 @@ def _synchronize_so_line_values(self, so_line):
             }
         return {}
 
+    def _sale_order_ticket_type_change_notify(self, new_ticket_id):
```
Comment (awa-odoo) [inline addons/event_sale/models/event_registration.py]: It's not an id, but an actual event_ticket.

Rename to `new_event_ticket`

File: addons/event_sale/models/event_registration.py [hunk 35]
```diff
@@ -60,6 +74,19 @@ def _synchronize_so_line_values(self, so_line):
...
+        for registration in self:
+            render_context = {
+                'registration': registration,
+                'old_ticket': registration.event_ticket_id.name,
+                'new_ticket': new_ticket_id.name
```
Comment (awa-odoo) [inline addons/event_sale/models/event_registration.py]: old_ticket_name
new_ticket_name

To make it clear these are not browserecords.

File: addons/event_sale/models/event_registration.py [hunk 36]
```diff
@@ -60,6 +74,19 @@ def _synchronize_so_line_values(self, so_line):
...
+            }
+            registration.sale_order_id.activity_schedule_with_view('mail.mail_activity_data_warning',
+                 user_id=registration.event_id.user_id.id or registration.sale_order_id.user_id.id or
+                         self.env.user.id if not self.env.user._is_public() else self.env.ref("base.user_admin").id,
```
Comment (awa-odoo) [inline addons/event_sale/models/event_registration.py]: Wow.

Make a bunch of if/elif above to determine a user_id instead to ease readability.

File: addons/event_sale/models/sale_order.py [hunk 37]
```diff
@@ -32,6 +34,24 @@ def action_confirm(self):
...
+            "view_mode": "tree,kanban,form",
+            "domain": [('sale_order_id', 'in', self.ids)]
+        }
+
+    def _compute_attendee_count(self):
+        sale_orders_data = self.env['event.registration'].read_group(
+            [('sale_order_id', 'in', self.ids)],
+            ['sale_order_id'], ['sale_order_id']
+        )
+        attendee_count_data = {sale_order_data['sale_order_id'][0]: sale_order_data['sale_order_id_count'] for sale_order_data in sale_orders_data}
```
Comment (awa-odoo) [inline addons/event_sale/models/sale_order.py]: Readability:
```
{
    sale_order_data['sale_order_id'][0]:
    sale_order_data['sale_order_id_count']
   for sale_order_data in sale_orders_data
}
```

File: addons/event_sale/models/sale_order.py [hunk 38]
```diff
@@ -32,6 +34,24 @@ def action_confirm(self):
...
         return res
 
+    def action_view_attendee_list(self):
+        return {
+            "name": _("Attendees"),
+            "type": "ir.actions.act_window",
+            "res_model": "event.registration",
+            "view_mode": "tree,kanban,form",
+            "domain": [('sale_order_id', 'in', self.ids)]
+        }
```
Comment (awa-odoo) [inline addons/event_sale/models/sale_order.py]: Can't use existing action?

File: addons/event_sale/models/event_registration.py [hunk 39]
```diff
@@ -25,6 +25,14 @@ def _check_auto_confirmation(self):
                 res = False
         return res
 
+    def action_view_sale_order(self):
+        return {
+            "type": "ir.actions.act_window",
+            "res_model": "sale.order",
+            "view_mode": "form",
+            "res_id": self.sale_order_id.id,
+        }
```
Comment (awa-odoo) [inline addons/event_sale/models/event_registration.py]: Can't use existing action?

File: addons/event_sale/views/event_registration_views.xml [hunk 40]
```diff
@@ -34,6 +34,12 @@
                         class="oe_stat_button" icon="fa-usd" string="Sale Order">
                 </button>
             </xpath>
+            <xpath expr="//div[@name='button_box']" position="after">
+                <widget name="web_ribbon" title="Paid" bg_color="bg-success" attrs="{'invisible': [('is_paid', '=', False)]}"/>
```
Comment (awa-odoo) [inline addons/event_sale/views/event_registration_views.xml]: Can you put the invisible field above your widget tag?
To avoid extra xpath?

File: addons/event_sale/models/event_registration.py [hunk 41]
```diff
@@ -60,21 +64,42 @@ def _synchronize_so_line_values(self, so_line):
...
-        information.append((_('Name'), self.name))
-        information.append((_('Ticket'), self.event_ticket_id.name or _('None')))
+    def _sale_order_ticket_type_change_notify(self, new_event_ticket):
+        for registration in self:
+            render_context = {
+                'registration': registration,
+                'old_ticket_name': registration.event_ticket_id.name,
+                'new_ticket_name': new_event_ticket.name
+            }
+            if registration.event_id.user_id:
```
Comment (tivisse) [inline addons/event_sale/models/event_registration.py]:
```suggestion
# Before loop
fallback_user_id = self.env.user.id if not self.env.user._is_public() else self.env.ref("base.user_admin").id
# In loop
user_id = registration.event_id.user_id.id or \
          registration.sale_order_id.user_id.id or \
          fallback_user_id
```

File: addons/website_event_track/controllers/main.py [hunk 42]
```diff
@@ -129,7 +131,7 @@ def event_track_proposal(self, event, **post):
         if not event.can_access_from_current_website():
             raise NotFound()
 
-        return request.render("website_event_track.event_track_proposal", {'event': event})
+        return request.render("website_event_track.event_track_proposal", {'event': event, 'main_object': event})
```
Comment (tivisse) [inline addons/website_event_track/controllers/main.py]: This doesn't seem to be necessary right ?

File: addons/website_event_track/controllers/main.py [hunk 43]
```diff
@@ -146,11 +148,12 @@ def event_track_proposal_post(self, event, **post):
             'partner_name': post['partner_name'],
             'partner_email': post['email_from'],
             'partner_phone': post['phone'],
-            'partner_biography': escape(post['biography']),
+            'partner_biography': post['biography'],
```
Comment (tivisse) [inline addons/website_event_track/controllers/main.py]: Should be sanitized at least. This is a public controller, so it shouldn't be trusted.

File: addons/website_event_track/controllers/main.py [hunk 44]
```diff
@@ -146,11 +148,12 @@ def event_track_proposal_post(self, event, **post):
...
             'event_id': event.id,
             'tag_ids': [(6, 0, tags)],
             'user_id': False,
-            'description': escape(post['description'])
+            'description': post['description'],
```
Comment (tivisse) [inline addons/website_event_track/controllers/main.py]: Same here.

File: addons/website_event_track/controllers/main.py [hunk 45]
```diff
@@ -146,11 +148,12 @@ def event_track_proposal_post(self, event, **post):
...
+            'image': base64.b64encode(post['image'].read()) if post.get('image') else False
```
Comment (tivisse) [inline addons/website_event_track/controllers/main.py]: We should handle the case when the image content is crap no?

File: addons/website_event_track/controllers/main.py (hunk 42 above)
Comment (pko-odoo) [reply addons/website_event_track/controllers/main.py]: Without it the tab name isn't set as we want unfortunately

File: addons/website_event_track/controllers/main.py (hunk 45 above)
Comment (pko-odoo) [reply addons/website_event_track/controllers/main.py]: Isn't the 'accept' attribute from the input sufficient ? It's currently set to accept all images

File: addons/event_sale/models/event_registration.py [hunk 46]
```diff
@@ -46,6 +45,11 @@ def write(self, vals):
                 self.env['sale.order.line'].browse(vals['sale_order_line_id'])
             )
             vals.update(so_line_vals)
+
+        if vals.get('event_ticket_id'):
+            if self.event_ticket_id and self.event_ticket_id.id != vals['event_ticket_id']:
+                self._sale_order_ticket_type_change_notify(self.env['event.event.ticket'].browse(vals['event_ticket_id']))
```
Comment (awa-odoo) [inline addons/event_sale/models/event_registration.py]:
```py
self.filtered(
    lambda registration: registration.event_ticket_id and registration.event_ticket_id.id != vals['event_ticket_id'])
)._sale_order_ticket_type_change_notify(self.env['event.event.ticket'].browse(vals['event_ticket_id']))
```

**Comment by tivisse** [comment]
@robodoo r+ merge

**Comment by tivisse** [comment]
@robodoo rebase-merge
